### PR TITLE
Fix nested DEF/USE nodes

### PR DESCRIPTION
**Pull request by stefaniapedrazzi**
**Description**
If USE node B has a DEF ancestor A, then USE node A is not updated when DEF node B changes.
Sample world:
```
Solid {
  children [
    Transform {
      children [
        DEF B Shape {
          appearance Appearance {
            material Material {
            }
          }
          geometry Box {
            size 0.1 0.1 0.1
        }
      ]
    }
    DEF A Transform {
      translation 0.4 0 0
      children [
        USE B
      ]
    }
  ]
  boundingObject USE A
}
```

**Related Issues**
This pull-request fixes issue #1538.

**Tasks**
* [x] Update DEF ancestor of USE nodes on field change
* [x] Check if USE node is up-to-date before forwarding the field changes
* [x] Add test

Second task fixes a crash when inserting a Material node into the following DEF A Appearance:
```
DEF G Group {
  children [
    DEF S Shape {
      appearance DEF A Appearance {
      }
    }
    USE S
  ]
}
USE G
```

### CODE REVIEW

**Comment by stefaniapedrazzi** [comment]
The original issue described in #1538 is fixed, i.e. when modifying an existing node.
But there are also issues when inserting/deleting nodes.

**Comment by stefaniapedrazzi** [comment]
:heavy_check_mark: travis_unit

**Comment by DavidMansolino** [comment]
:eyes:

File: tests/api/worlds/nested_def_nodes.wbt [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+              }
+            }
+          ]
+        }
+        DEF T2 Transform {
+          translation 0.4 0 0
+          children [
+            USE S1
+          ]
+        }
```
Comment (DavidMansolino) [inline tests/api/worlds/nested_def_nodes.wbt:44]: Adding a `USE T2` here causes an issue:
  - changing the `B1` size works fine.
  - changing the `T2` translation works fine for the shape but not for the bounding object (a save and revert is still required for the update to be effective).

**Comment by stefaniapedrazzi** [comment]
Thank you for testing. I will check these cases too.

**Comment by stefaniapedrazzi** [comment]
The cause of the issue with the Transform.translation change seems the same as the one described in #1569: the field redirection is not correct for USE nodes and it doesn't trigger the update of the USE nodes connected to the upper DEF.
I will review the node creation and field redirection to try to properly fix also these cases.

File: tests/api/worlds/nested_def_nodes.wbt (hunk 1 above)
Comment (stefaniapedrazzi) [reply tests/api/worlds/nested_def_nodes.wbt:44]: I added a test for this case below (`USE_CASE_4`): test the update of a USE node containing another USE node.

**Comment by stefaniapedrazzi** [comment]
:heavy_check_mark: travis_unit

Issues reported just above by @DavidMansolino and #1569 are now fixed.

I checked if the issues come from the node instantiation, but there is no field redirection specific to DEF/USE nodes in the node constructor. So it is the `WbNode::notifyFieldChanged` function that have to take care of it.

**Comment by DavidMansolino** [comment]
👀

**Pull request merged**